schematic/schemas/json_schema_generator.py

```diff
@@ -0,0 +1,747 @@
+"Data Model Json Schema"
+
+import logging
+import os
+from typing import Union, Any
+from dataclasses import dataclass, field, asdict
+import networkx as nx  # type: ignore
+
+from schematic.schemas.data_model_graph import DataModelGraphExplorer
+from schematic.schemas.data_model_relationships import DataModelRelationships
+from schematic.utils.schema_utils import get_json_schema_log_file_path
+from schematic.utils.validate_utils import rule_in_rule_list
+
+from schematic.utils.io_utils import export_json
+
+
+logger = logging.getLogger(__name__)
+
+# A dict where the keys are type validation rules, and the values are their JSON Schema equivalent
+TYPE_RULES = {
+    "str": "string",
+    "num": "number",
+    "float": "number",
+    "int": "integer",
+    "bool": "boolean",
+}
+
+
+@dataclass
+class PropertyData:
+    """A Dataclass representing data about a JSON Schema property from its validation rules"""
+
+    property_type: Union[str, None] = None
+    is_array: bool = False
+    range_min: Union[float, None] = None
+    range_max: Union[float, None] = None
+
+
+@dataclass
+class JSONSchema:  # pylint: disable=too-many-instance-attributes
+    """A dataclass representing a JSON Schema"""
+
+    schema_id: str = ""
+    title: str = ""
+    schema: str = "http://json-schema.org/draft-07/schema#"
+    type: str = "object"
+    description: str = "TBD"
+    properties: dict[str, Any] = field(default_factory=dict)
+    required: list[str] = field(default_factory=list)
+    all_of: list[dict[str, Any]] = field(default_factory=list)
+
+    def as_json_schema_dict(self) -> dict[str, Any]:
+        """Returns class as a JSON Schema dictionary, with proper keywords"""
+        json_schema_dict = asdict(self)
+        keywords_to_change = {
+            "schema_id": "$id",
+            "schema": "$schema",
+            "all_of": "allOf",
+        }
+        for old_word, new_word in keywords_to_change.items():
+            json_schema_dict[new_word] = json_schema_dict.pop(old_word)
+        if len(self.all_of) == 0:
+            json_schema_dict.pop("allOf")
+        return json_schema_dict
+
+
+class NodeProcessor:
```

Comment:
nit: Unless this cannot be a dataclass I'd prefer to keep consistent behavior for classes introduced in this code.